### DISCUSSION TITLE
Welcome to bpftime Discussions!

### DISCUSSION THREAD

**Discussion by yunwei37** [eunomia-bpf/bpftime#64] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by venumadhavjosyula** [comment]
I am getting following error
[vjosyula@localhost bpftime]$ make build
cmake -Bbuild -DBPFTIME_ENABLE_UNIT_TESTING=1 -DBUILD_BPFTIME_DAEMON=1 -DCMAKE_BUILD_TYPE:STRING=Debug
-- The C compiler identification is GNU 11.4.1
-- The CXX compiler identification is GNU 11.4.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Enabling ubsan for Debug builds; Processor=x86_64
-- Started CMake for bpftime v0.1.0...

CMake Error at /usr/share/cmake/Modules/ExternalProject.cmake:1721 (file):
  file problem creating directory:
  /home/vjosyula/work/bpftime_enunomia/building_from_source/bpftime/third_party/libbpf//src
Call Stack (most recent call first):
  /usr/share/cmake/Modules/ExternalProject.cmake:3633 (_ep_set_directories)
  cmake/libbpf.cmake:6 (ExternalProject_Add)
  CMakeLists.txt:67 (include)


CMake Error at /usr/share/cmake/Modules/ExternalProject.cmake:1723 (message):
  dir
  '/home/vjosyula/work/bpftime_enunomia/building_from_source/bpftime/third_party/libbpf//src'
  does not exist after file(MAKE_DIRECTORY)
Call Stack (most recent call first):
  /usr/share/cmake/Modules/ExternalProject.cmake:3633 (_ep_set_directories)
  cmake/libbpf.cmake:6 (ExternalProject_Add)
  CMakeLists.txt:67 (include)


-- Configuring incomplete, errors occurred!
See also "/home/vjosyula/work/bpftime_enunomia/building_from_source/bpftime/build/CMakeFiles/CMakeOutput.log".
make: *** [Makefile:50: build] Error 1
[vjosyula@localhost bpftime]$ 

commands used are
make build

[vjosyula@localhost bpftime]$ pwd
/home/vjosyula/work/bpftime_enunomia/building_from_source/bpftime
[vjosyula@localhost bpftime]$

**Comment by yunwei37** [comment]
hi @venumadhavjosyula , have you run `git submodule update --init --recursive`?

Please refer to https://eunomia.dev/bpftime/documents/build-and-test/#table-of-contents for details

**Reply by venumadhavjosyula** [reply]
@yunwei37 after doing you suggested I was able to do needful. Thanks , for your help

**Comment by venumadhavjosyula** [comment]
Another question, I trying to attach random process in my case it is firefox process
[root@localhost bpftime]# bpftime attach 145025
[2024-03-01 02:35:36.276] [info] Injecting to 145025
[2024-03-01 02:35:36.351] [error] Failed to inject: Bootstrapper failed due to 'auxv-not-found'; please file a bug


Ps dump

root        7398    7394  0 Feb29 ?        00:00:00 fusermount -o rw,nosuid,nodev,fsname=portal,auto_unmount,subtype=portal -- /run/user/1000/doc
vjosyula    7404    5449  0 Feb29 ?        00:00:02 /usr/libexec/xdg-desktop-portal-gnome
vjosyula    7418    5449  0 Feb29 ?        00:00:02 /usr/libexec/xdg-desktop-portal-gtk
root      135085       1  0 Feb29 ?        00:00:35 /usr/bin/containerd
root      135103       1  0 Feb29 ?        00:01:09 /usr/bin/dockerd -H fd:// --containerd=/run/containerd/containerd.sock
root      142502       2  0 Feb29 ?        00:00:00 [kworker/7:0-xfs-inodegc/dm-0]
vjosyula  144928    5449  0 Feb29 ?        00:00:29 /usr/lib64/firefox/firefox --new-window
vjosyula  145025  144928  0 Feb29 ?        00:00:00 /usr/lib64/firefox/firefox -contentproc -parentBuildID 20240130173142 -prefsLen 28482 -prefMapSize 236314 -appDir /usr/lib64/firefox/brow
vjosyula  145072  144928  0 Feb29 ?        00:00:01 /usr/lib64/firefox/firefox -contentproc -childID 1 -isForBrowser -prefsLen 28623 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202
vjosyula  145124  144928  0 Feb29 ?        00:00:00 /usr/lib64/firefox/firefox -contentproc -childID 2 -isForBrowser -prefsLen 34102 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202
vjosyula  145191  144928  0 Feb29 ?        00:00:04 /usr/lib64/firefox/firefox -contentproc -childID 3 -isForBrowser -prefsLen 29235 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202
vjosyula  145193  144928  0 Feb29 ?        00:00:00 /usr/lib64/firefox/firefox -contentproc -childID 4 -isForBrowser -prefsLen 29235 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202
vjosyula  145234  144928  0 Feb29 ?        00:00:00 /usr/lib64/firefox/firefox -contentproc -childID 5 -isForBrowser -prefsLen 29235 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202
vjosyula  145324  144928  0 Feb29 ?        00:00:00 /usr/lib64/firefox/firefox -contentproc -childID 7 -isForBrowser -prefsLen 29373 -prefMapSize 236314 -jsInitLen 240916 -parentBuildID 202

Am I missing something.

**Comment by yunwei37** [comment]
Seems like an error related to Frida, you can read the tools/cli-cpp source code, make a small poc and report the error to https://github.com/frida/frida-gum

**Comment by venumadhavjosyula** [comment]
Totally different no related to above

We are using example/minimal/uprobe.bpf.c and example/minimal/uprobe.c for our use-case

In the code below
SEC("uprobe//home/vjosyula/work/bpftime_enunomia/building_from_source/dpdk_examples/The-DPDK-Examples/build/simple_l3fwd:fwd_pckt")
int do_uprobe_trace(struct pt_regs *ctx)
{
unsigned port_id;
bpf_probe_read(&port_id, sizeof(unsigned ), (void *)PT_REGS_PARM1(ctx));
bpf_printk("target_func called. Ox%x\n",port_id);
return 0;
}

The definition of fwd_pkt is as follows,
void fwd_pckt(struct rte_mbuf *pckt, unsigned port_id, struct rte_hash *route_tbl)

In in do_uprobe_trace , how do I access these three i/p params using struct pt_regs *ctx ?

Can you please suggest. I want to access i/p params of function being uprobed using *ctx

**Comment by lkyptw** [comment]
Hi, I'm study bpftime recently. I have a question that does bpftime can use on ARM architecture machine? If no, and why? Thanks.

**Reply by Officeyutong** [reply]
uprobe related features should work on ARM, but we have few tries

**Comment by haaHh** [comment]
Hi, I followed the example of minimal and found that running ebpf program alone can hook the target function, but using the following example did not achieve the desired effect.
```
code/bpftime/example/minimal# LD_PRELOAD=~/.bpftime/libbpftime-syscall-server.so ./uprobe
or
code/bpftime/example/minimal# bpftime load ./uprobe
llvm register vm factory
[2025-04-08 10:24:58.457] [info] [bpftime_config.cpp:81] Using VM: llvm
[2025-04-08 10:24:58.457] [info] [syscall_context.cpp:83] Init bpftime syscall mocking..
[2025-04-08 10:24:58.457] [info] [syscall_context.cpp:85] The log will be written to: ~/.bpftime/runtime.log
libbpf: loading object 'uprobe_bpf' from buffer
libbpf: elf: section(2) .symtab, size 120, link 1, flags 0, type=2
libbpf: elf: section(3) uprobe/./victim:target_func, size 128, link 0, flags 6, type=1
libbpf: sec 'uprobe/./victim:target_func': found program 'do_uprobe_trace' at insn offset 0 (0 bytes), code size 16 insns (128 bytes)
libbpf: elf: section(4) .rodata.str1.1, size 21, link 0, flags 32, type=1
libbpf: elf: section(5) license, size 4, link 0, flags 3, type=1
libbpf: license of uprobe_bpf is GPL
libbpf: elf: section(6) .BTF, size 801, link 0, flags 0, type=1
libbpf: elf: section(7) .BTF.ext, size 112, link 0, flags 0, type=1
libbpf: looking for externs among 5 symbols...
libbpf: collected 0 externs total
libbpf: map '.rodata.str1.1' (global data): at sec_idx 4, offset 0, flags 80.
[2025-04-08 10:24:58.459] [info] [bpftime_config.cpp:81] Using VM: llvm
libbpf: map 0 is ".rodata.str1.1"
libbpf: map '.rodata.str1.1': created successfully, fd=5
libbpf: elf: symbol address match for 'target_func' in './victim': 0x11a9
```

How can I troubleshoot the problem?

**Reply by haaHh** [reply]
env: Ubuntu 23.04

**Reply by Officeyutong** [reply]
> Hi, I followed the example of minimal and found that running ebpf program alone can hook the target function, but using the following example did not achieve the desired effect.
> 
> ```
> code/bpftime/example/minimal# LD_PRELOAD=~/.bpftime/libbpftime-syscall-server.so ./uprobe
> or
> code/bpftime/example/minimal# bpftime load ./uprobe
> llvm register vm factory
> [2025-04-08 10:24:58.457] [info] [bpftime_config.cpp:81] Using VM: llvm
> [2025-04-08 10:24:58.457] [info] [syscall_context.cpp:83] Init bpftime syscall mocking..
> [2025-04-08 10:24:58.457] [info] [syscall_context.cpp:85] The log will be written to: ~/.bpftime/runtime.log
> libbpf: loading object 'uprobe_bpf' from buffer
> libbpf: elf: section(2) .symtab, size 120, link 1, flags 0, type=2
> libbpf: elf: section(3) uprobe/./victim:target_func, size 128, link 0, flags 6, type=1
> libbpf: sec 'uprobe/./victim:target_func': found program 'do_uprobe_trace' at insn offset 0 (0 bytes), code size 16 insns (128 bytes)
> libbpf: elf: section(4) .rodata.str1.1, size 21, link 0, flags 32, type=1
> libbpf: elf: section(5) license, size 4, link 0, flags 3, type=1
> libbpf: license of uprobe_bpf is GPL
> libbpf: elf: section(6) .BTF, size 801, link 0, flags 0, type=1
> libbpf: elf: section(7) .BTF.ext, size 112, link 0, flags 0, type=1
> libbpf: looking for externs among 5 symbols...
> libbpf: collected 0 externs total
> libbpf: map '.rodata.str1.1' (global data): at sec_idx 4, offset 0, flags 80.
> [2025-04-08 10:24:58.459] [info] [bpftime_config.cpp:81] Using VM: llvm
> libbpf: map 0 is ".rodata.str1.1"
> libbpf: map '.rodata.str1.1': created successfully, fd=5
> libbpf: elf: symbol address match for 'target_func' in './victim': 0x11a9
> ```
> 
> How can I troubleshoot the problem?

Have you executed `bpftime start ./victim` or something alike?

**Reply by haaHh** [reply]
yes

code/bpftime/example/minimal# bpftime start ./victim
llvm register vm factory
[2025-04-09 01:52:56.494] [info] [bpftime_config.cpp:81] Using VM: llvm
llvm register vm factory
[2025-04-09 01:52:56.501] [info] [bpftime_config.cpp:81] Using VM: llvm
[2025-04-09 01:52:56.501] [info] [bpftime_shm_internal.cpp:673] Global shm constructed. shm_open_type 1 for bpftime_maps_shm
[2025-04-09 01:52:56.501] [info] [bpftime_shm_internal.cpp:34] Global shm initialized
target_func
target_func
target_func


tail -f  ~/.bpftime/runtime.log

[2025-04-09 01:54:55][info][261] Register attach-impl defined helper bpf_get_func_arg, index 183
[2025-04-09 01:54:55][info][261] Register attach-impl defined helper bpf_get_func_ret_id, index 184
[2025-04-09 01:54:55][info][261] Register attach-impl defined helper bpf_get_retval, index 186
[2025-04-09 01:54:55][info][261] Initializing agent..
[2025-04-09 01:54:55][info][261] Attach successfully
[2025-04-09 01:54:59][info][263] Initialize syscall server
[2025-04-09 01:54:59][info][263] Using VM: llvm
[2025-04-09 01:54:59][info][263] Global shm constructed. shm_open_type 0 for bpftime_maps_shm
[2025-04-09 01:54:59][info][263] Global shm initialized
[2025-04-09 01:54:59][info][263] bpftime-syscall-server started
[2025-04-09 01:54:59][info][263] Created uprobe/uretprobe perf event handler, module name ./victim, offset 11a9

**Reply by haaHh** [reply]
Hook was successfully combined with the following commands.

code/bpftime/example/minimal# ./uprobe

and

#code/bpftime/example/minimal# bpftime start ./victim or
code/bpftime/example/minimal# LD_PRELOAD=~/.bpftime/libbpftime-agent.so ./victim

Then you can see print in the directory of /sys/kernel/debug/tracing/trace_pipe.

**Reply by Officeyutong** [reply]
> Hook was successfully combined with the following commands.
> 
> code/bpftime/example/minimal# ./uprobe
> 
> and
> 
> #code/bpftime/example/minimal# bpftime start ./victim or code/bpftime/example/minimal# LD_PRELOAD=~/.bpftime/libbpftime-agent.so ./victim
> 
> Then you can see print in the directory of /sys/kernel/debug/tracing/trace_pipe.

To use bpftime, you should run `bpftime load ./uprobe` in one terminal, and `bpftime start ./victim` on another terminal. Note that `bpftime load ./uprobe` (a.k.a syscall server) must keep alive during `bpftime start ./victim` (a.k.a agent)

**Reply by haaHh** [reply]
> > Hook was successfully combined with the following commands.
> > code/bpftime/example/minimal# ./uprobe
> > and
> > #code/bpftime/example/minimal# bpftime start ./victim or code/bpftime/example/minimal# LD_PRELOAD=~/.bpftime/libbpftime-agent.so ./victim
> > Then you can see print in the directory of /sys/kernel/debug/tracing/trace_pipe.
> 
> To use bpftime, you should run `bpftime load ./uprobe` in one terminal, and `bpftime start ./victim` on another terminal. Note that `bpftime load ./uprobe` (a.k.a syscall server) must keep alive during `bpftime start ./victim` (a.k.a agent)

![17441921456321](https://github.com/user-attachments/assets/a6fd2a09-822b-4aa4-bf33-40ba62a2f1b6)

I ran like this, but I didn't see any printf

**Reply by haaHh** [reply]
ex:
The time spent running benchmark/uprobe/uprobe with bpftime is the same as the time spent running empty cases, but the running time with kernel will become very large.

bpftime run benchmark/uprobe/uprobe
![bpftime_run](https://github.com/user-attachments/assets/fd22d3d6-d7ec-44bb-85f1-64a465d8c724)

run benchmark test alone
![empty](https://github.com/user-attachments/assets/e827e7ec-d2d0-45eb-a294-30d22f1a1a0d)

without bpftime
![image](https://github.com/user-attachments/assets/937fff6d-9a01-4c11-8e1d-950fd7ddd033)

**Comment by haoowa** [comment]
Hi, how can I attach a CUDA process to profile the details？my steps are below：
1. Run vector_add in background, `./examples/gpu/kernelretsnoop/vector_add 2>&1 > /dev/null &`.
2. Open an another terminal to load the ebpf, `BPFTIME_SHM_MEMORY_MB=100 bpftime load ./examples/gpu/kernelretsnoop/kernelretsnoop`
<img width="1270" height="765" alt="image" src="https://github.com/user-attachments/assets/ee83e764-5058-4e92-ab05-452ff3b115c8" />
3. Do attach, it succeed, but nothing happens in terminal 2. 
<img width="853" height="656" alt="image" src="https://github.com/user-attachments/assets/5ca97918-03a4-407e-a8ec-3d18d1074aab" />

Wasn't the program supposed to print something?
